engine/access/rpc/backend/transaction_subscription_metadata.go

```diff
@@ -0,0 +1,255 @@
+package backend
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/onflow/flow-go/engine/access/subscription"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/onflow/flow-go/access"
+	"github.com/onflow/flow-go/engine/common/rpc"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/irrecoverable"
+	"github.com/onflow/flow-go/state"
+	"github.com/onflow/flow-go/storage"
+
+	"github.com/onflow/flow/protobuf/go/flow/entities"
+)
+
+// transactionSubscriptionMetadata manages the state of a transaction subscription.
+//
+// This struct contains metadata for tracking a transaction's progress, including
+// references to relevant blocks, collections, and transaction results.
+type transactionSubscriptionMetadata struct {
+	blocks               storage.Blocks
+	collections          storage.Collections
+	transactions         storage.Transactions
+	txResult             *access.TransactionResult
+	txReferenceBlockID   flow.Identifier
+	blockWithTx          *flow.Header
+	eventEncodingVersion entities.EventEncodingVersion
+
+	backendTransactions *backendTransactions
+}
+
+// newTransactionSubscriptionMetadata initializes a new metadata object for a transaction subscription.
+//
+// This function constructs a transaction metadata object used for tracking the transaction's progress
+// and maintaining its state throughout execution.
+//
+// Parameters:
+//   - ctx: Context for managing the lifecycle of the operation.
+//   - backendTransactions: A reference to the backend transaction manager.
+//   - txID: The unique identifier of the transaction.
+//   - txReferenceBlockID: The ID of the transaction’s reference block.
+//   - eventEncodingVersion: The required version of event encoding.
+//
+// Returns:
+//   - *transactionSubscriptionMetadata: The initialized transaction metadata object.
+//
+// No errors expected during normal operations.
+func newTransactionSubscriptionMetadata(
+	ctx context.Context,
+	backendTransactions *backendTransactions,
+	txID flow.Identifier,
+	txReferenceBlockID flow.Identifier,
+	eventEncodingVersion entities.EventEncodingVersion,
+) (*transactionSubscriptionMetadata, error) {
+	txMetadata := &transactionSubscriptionMetadata{
+		backendTransactions:  backendTransactions,
+		txResult:             &access.TransactionResult{TransactionID: txID},
+		eventEncodingVersion: eventEncodingVersion,
+		blocks:               backendTransactions.blocks,
+		collections:          backendTransactions.collections,
+		transactions:         backendTransactions.transactions,
+		txReferenceBlockID:   txReferenceBlockID,
+	}
+
+	if err := txMetadata.Refresh(ctx); err != nil {
```

Comment:
does this need to be called here? Is it safe to wait until the first call in `getTransactionStatusResponse()`? ideally there is no logic in the constructor unless totally necessary